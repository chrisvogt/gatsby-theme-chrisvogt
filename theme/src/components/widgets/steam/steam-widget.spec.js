import React from 'react'
import renderer, { act } from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import { configureStore as configureRealStore } from '@reduxjs/toolkit'
// import SteamWidget from './steam-widget' // REMOVE THIS LINE

// Mock child components to isolate the test
jest.mock('../call-to-action', () => props => <div data-testid='CallToAction'>{props.title}</div>)
jest.mock('../recent-posts/post-card', () => props => <div data-testid='PostCard'>{props.title}</div>)
jest.mock('../profile-metrics-badge', () => props => (
  <div data-testid='ProfileMetricsBadge'>{JSON.stringify(props.metrics)}</div>
))
jest.mock('../widget', () => props => <div data-testid='Widget'>{props.children}</div>)
jest.mock('../widget-header', () => props => <div data-testid='WidgetHeader'>{props.children}</div>)

// Mock hooks and selectors
jest.mock('../../../hooks/use-site-metadata', () => () => ({
  widgets: { steam: { url: 'https://example.com/steam-feed' } }
}))
jest.mock('../../../selectors/metadata', () => ({
  getSteamWidgetDataSource: () => 'https://example.com/steam-feed'
}))

// Move fetchDataSource mock before importing the component
let mockFetchDataSource
jest.mock('../../../actions/fetchDataSource', () => {
  mockFetchDataSource = jest.fn(() => ({ type: 'MOCK_ACTION' }))
  return mockFetchDataSource
})

const mockStore = configureMockStore([]) // no middleware

// Re-import after mocks
const SteamWidget = require('./steam-widget').default

describe('SteamWidget', () => {
  it('renders correctly with sample data', () => {
    const store = mockStore({
      widgets: {
        steam: {
          state: 'SUCCESS',
          data: {
            metrics: [{ label: 'Games Played', value: 5 }],
            profile: {
              displayName: 'Chris',
              profileURL: 'https://steamcommunity.com/id/themeuser'
            },
            collections: {
              recentlyPlayedGames: [
                {
                  id: '123',
                  displayName: 'Half-Life',
                  playTime2Weeks: 120,
                  images: { header: 'https://example.com/halflife.jpg' }
                },
                {
                  id: '456',
                  displayName: 'Portal',
                  playTime2Weeks: 45,
                  images: { header: 'https://example.com/portal.jpg' }
                }
              ],
              ownedGames: [
                {
                  id: '255710',
                  displayName: 'Cities: Skylines',
                  playTimeForever: 45441,
                  playTime2Weeks: 120,
                  images: { icon: 'https://example.com/cities-icon.jpg' }
                },
                {
                  id: '346110',
                  displayName: 'ARK: Survival Evolved',
                  playTimeForever: 16670,
                  playTime2Weeks: null,
                  images: { icon: 'https://example.com/ark-icon.jpg' }
                }
              ]
            }
          }
        }
      }
    })

    const tree = renderer
      .create(
        <Provider store={store}>
          <SteamWidget />
        </Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders loading state', () => {
    const store = mockStore({
      widgets: {
        steam: {
          state: 'LOADING'
        }
      }
    })

    const tree = renderer
      .create(
        <Provider store={store}>
          <SteamWidget />
        </Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders error state (hasFatalError)', () => {
    const store = mockStore({
      widgets: {
        steam: {
          state: 'SUCCESS',
          data: {
            metrics: [{ label: 'Games Played', value: 5 }],
            profile: {
              displayName: 'Chris',
              profileURL: 'https://steamcommunity.com/id/themeuser'
            },
            collections: {
              recentlyPlayedGames: [],
              ownedGames: []
            }
          },
          hasFatalError: true
        }
      }
    })
    const tree = renderer
      .create(
        <Provider store={store}>
          <SteamWidget />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders AI summary if present', () => {
    const store = mockStore({
      widgets: {
        steam: {
          state: 'SUCCESS',
          data: {
            aiSummary: 'This is an AI summary',
            metrics: [{ label: 'Games Played', value: 5 }],
            profile: {
              displayName: 'Chris',
              profileURL: 'https://steamcommunity.com/id/themeuser'
            },
            collections: {
              recentlyPlayedGames: [],
              ownedGames: []
            }
          }
        }
      }
    })
    const tree = renderer
      .create(
        <Provider store={store}>
          <SteamWidget />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders with empty metrics', () => {
    const store = mockStore({
      widgets: {
        steam: {
          state: 'SUCCESS',
          data: {
            metrics: [],
            profile: {
              displayName: 'Chris',
              profileURL: 'https://steamcommunity.com/id/themeuser'
            },
            collections: {
              recentlyPlayedGames: [],
              ownedGames: []
            }
          }
        }
      }
    })
    const tree = renderer
      .create(
        <Provider store={store}>
          <SteamWidget />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders with empty recentlyPlayedGames and ownedGames', () => {
    const store = mockStore({
      widgets: {
        steam: {
          state: 'SUCCESS',
          data: {
            metrics: [{ label: 'Games Played', value: 5 }],
            profile: {
              displayName: 'Chris',
              profileURL: 'https://steamcommunity.com/id/themeuser'
            },
            collections: {
              recentlyPlayedGames: [],
              ownedGames: []
            }
          }
        }
      }
    })
    const tree = renderer
      .create(
        <Provider store={store}>
          <SteamWidget />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls fetchDataSource when isLoading is true', async () => {
    mockFetchDataSource.mockClear()
    // Use a real Redux store for this test
    const initialState = {
      widgets: {
        steam: {
          state: 'INIT',
          data: {
            metrics: [],
            profile: {},
            collections: { recentlyPlayedGames: [], ownedGames: [] }
          }
        }
      }
    }
    // Minimal reducer to support the selectors
    function widgets(state = initialState.widgets) {
      return state
    }
    const store = configureRealStore({ reducer: { widgets } })
    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <SteamWidget />
        </Provider>
      )
    })
    expect(mockFetchDataSource).toHaveBeenCalled()
  })
})
