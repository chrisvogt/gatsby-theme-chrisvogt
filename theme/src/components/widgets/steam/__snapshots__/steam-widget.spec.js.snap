// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SteamWidget renders correctly with sample data 1`] = `
<div
  data-testid="Widget"
>
  <div
    data-testid="WidgetHeader"
  >
    Steam
  </div>
  <div
    data-testid="ProfileMetricsBadge"
  >
    [{"label":"Games Played","value":5}]
  </div>
  <div
    className="css-gavibk"
  >
    <h3
      className="css-1pgqsji"
    >
      My Games
    </h3>
  </div>
  <p
    className="css-1v9l0er"
  >
    Games I own and their play time statistics.
  </p>
  <div
    data-testid="OwnedGamesTable"
  >
    [{"id":"255710","displayName":"Cities: Skylines","playTimeForever":45441,"playTime2Weeks":120,"images":{"icon":"https://example.com/cities-icon.jpg"}},{"id":"346110","displayName":"ARK: Survival Evolved","playTimeForever":16670,"playTime2Weeks":null,"images":{"icon":"https://example.com/ark-icon.jpg"}}]
  </div>
  <div
    className="css-19lpcf0"
  >
    <h3
      className="css-1pgqsji"
    >
      Recently-Played Games
    </h3>
  </div>
  <p
    className="css-1v9l0er"
  >
    Games I've played in the last two weeks.
  </p>
  <div
    className="css-1bz685e"
  >
    <div
      data-testid="PostCard"
    >
      Half-Life
    </div>
    <div
      data-testid="PostCard"
    >
      Portal
    </div>
  </div>
</div>
`;

exports[`SteamWidget renders loading state 1`] = `
<div
  data-testid="Widget"
>
  <div
    data-testid="WidgetHeader"
  >
    Steam
  </div>
  <div
    data-testid="ProfileMetricsBadge"
  >
    []
  </div>
  <div
    className="css-gavibk"
  >
    <h3
      className="css-1pgqsji"
    >
      My Games
    </h3>
  </div>
  <p
    className="css-1v9l0er"
  >
    Games I own and their play time statistics.
  </p>
  <div
    data-testid="OwnedGamesTable"
  >
    []
  </div>
  <div
    className="css-19lpcf0"
  >
    <h3
      className="css-1pgqsji"
    >
      Recently-Played Games
    </h3>
  </div>
  <p
    className="css-1v9l0er"
  >
    Games I've played in the last two weeks.
  </p>
  <div
    className="css-1bz685e"
  />
</div>
`;

exports[`TimeSpent renders humanized time correctly 1`] = `"0 hours"`;
